pkg/controller/constraint/constraint_controller.go

```diff
@@ -217,40 +240,30 @@ func (r *ReconcileConstraint) Reconcile(request reconcile.Request) (reconcile.Re
 	}()
 
 	if HasFinalizer(instance) {
-		status, _, _ := unstructured.NestedFieldCopy(instance.Object, "status")
 		RemoveFinalizer(instance)
 		if err := r.writer.Update(context.Background(), instance); err != nil {
 			return reconcile.Result{Requeue: true}, nil
 		}
-
-		if status != nil {
-			if err := unstructured.SetNestedField(instance.Object, status, "status"); err != nil {
-				log.Error(err, "error preserving constraint status")
-			}
-		}
 	}
 
 	if !deleted {
 		r.log.Info("handling constraint update", "instance", instance)
-		status, err := csutil.GetHAStatus(instance)
+		status, err := r.getOrCreatePodStatus(instance)
 		if err != nil {
+			log.Info("could not get/create pod status object", "error", err)
 			return reconcile.Result{}, err
 		}
-		status.Errors = nil
-		if err = csutil.SetHAStatus(instance, status); err != nil {
-			return reconcile.Result{}, err
-		}
+		status.Status.ConstraintUID = instance.GetUID()
+		status.Status.ObservedGeneration = instance.GetGeneration()
+		status.Status.Errors = nil
 		if c, err := r.opa.GetConstraint(context.TODO(), instance); err != nil || !constraints.SemanticEqual(instance, c) {
 			if err := r.cacheConstraint(instance); err != nil {
 				r.constraintsCache.addConstraintKey(constraintKey, tags{
 					enforcementAction: enforcementAction,
 					status:            metrics.ErrorStatus,
 				})
-				status.Errors = append(status.Errors, csutil.Error{Message: err.Error()})
-				if err2 := csutil.SetHAStatus(instance, status); err2 != nil {
-					log.Error(err2, "could not set constraint error status")
-				}
-				if err2 := r.statusClient.Status().Update(context.TODO(), instance); err2 != nil {
+				status.Status.Errors = append(status.Status.Errors, statusv1beta1.Error{Message: err.Error()})
+				if err2 := r.writer.Update(context.TODO(), status); err2 != nil {
```

Comment:
Are we worried about concurrent status writers? I believe the answer is no, but I wanted to verify.

Reply:
Correct, the constraint/template controllers are the only controllers that write to the status object.

The only remaining collision is the audit thread is also writing to the status subresource of constraints/templates, but that conflict should be minimal and much less correlated as opposed to all controllers writing status on update.